flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/BufferedUpsertSink.java

```diff
@@ -0,0 +1,93 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.kafka.table;
+
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.connector.sink.Committer;
+import org.apache.flink.api.connector.sink.GlobalCommitter;
+import org.apache.flink.api.connector.sink.Sink;
+import org.apache.flink.api.connector.sink.SinkWriter;
+import org.apache.flink.core.io.SimpleVersionedSerializer;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.DataType;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * A wrapper of a {@link Sink}. It will buffer the data emitted by the wrapper {@link SinkWriter}
+ * and only emit it when the buffer is full or a timer is triggered or a checkpoint happens.
+ */
+class BufferedUpsertSink<WriterState> implements Sink<RowData, Void, WriterState, Void> {
+
+    private final Sink<RowData, ?, WriterState, ?> wrappedSink;
+    private final DataType physicalDataType;
+    private final int[] keyProjection;
+    private final SinkBufferFlushMode bufferFlushMode;
+    @Nullable private final TypeSerializer<RowData> typeSerializer;
+
+    BufferedUpsertSink(
+            Sink<RowData, ?, WriterState, ?> wrappedSink,
+            DataType physicalDataType,
+            int[] keyProjection,
+            SinkBufferFlushMode bufferFlushMode,
+            @Nullable TypeSerializer<RowData> typeSerializer) {
+        this.wrappedSink = wrappedSink;
+        this.physicalDataType = physicalDataType;
+        this.keyProjection = keyProjection;
+        this.bufferFlushMode = bufferFlushMode;
+        this.typeSerializer = typeSerializer;
+    }
+
+    @Override
+    public SinkWriter<RowData, Void, WriterState> createWriter(
+            InitContext context, List<WriterState> states) throws IOException {
+        final SinkWriter<RowData, ?, WriterState> wrapperWriter =
+                wrappedSink.createWriter(context, states);
+        return new BufferedUpsertWriter<>(
+                wrapperWriter, physicalDataType, keyProjection, bufferFlushMode, typeSerializer);
+    }
+
+    @Override
+    public Optional<SimpleVersionedSerializer<WriterState>> getWriterStateSerializer() {
+        return wrappedSink.getWriterStateSerializer();
+    }
+
+    @Override
+    public Optional<Committer<Void>> createCommitter() throws IOException {
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<GlobalCommitter<Void, Void>> createGlobalCommitter() throws IOException {
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<SimpleVersionedSerializer<Void>> getCommittableSerializer() {
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<SimpleVersionedSerializer<Void>> getGlobalCommittableSerializer() {
+        return Optional.empty();
```

Comment:
Delegate to wrapped sink!

Reply:
I've added a comment explaining why a proper two-phase commit is not needed because of the eventual consistency of the idempotent updates.